### PR TITLE
[BH-1406] Low battery turnon

### PR DESCRIPTION
**Pull request by paweljonskim**
Actions changes explaination:
Present state:
When an action is going to an application that is not turned on, system is launching the app and skipping the action. The action **will not be** processed until another action arrives.
Implemented change:
After application launch and switch confirmation, ApplicationManager processes all actions that are queued (skipped before) in the actionsRegistry

### CODE REVIEW

**Comment by paweljonskim** [comment]
https://user-images.githubusercontent.com/79840715/155357725-e9f3bb42-7649-455f-a2ac-24444b652bbd.mp4

The reset instead of shutdown in 30s is not connected to those changes, so dont mind that please :)

**Review by pholat — CHANGES_REQUESTED** [review]
please thoroughly check on pure as well as take into consideration action trigger placement

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp [hunk 1]
```diff
@@ -713,6 +717,7 @@ namespace app::manager
             setState(State::Running);
             EventManagerCommon::messageSetApplication(this, app.name());
             onLaunchFinished(app);
+            actionsRegistry.process();
```
[the remark this answers is not in the document]
Comment (paweljonskim) [reply module-services/service-appmgr/model/ApplicationManagerCommon.cpp:720]: alright, so lets dive deeper.
It might be confusing, but `hasPendingAction` does not mean "is there an action in the queue we need to process" but rather "is there an action going on at the moment, which we are waiting to finish". And guess what, there isnt in this case.
That mechanism of launching an app where there is an action incoming **does not use Launch action**. During the application launch caused by incoming custom action **there is no "pending action"**. One might ask "shouldnt there be a pending action then?" but I insist on not reworking the whole action subsystem during emergency pre-release bugfix :)

So basically I agree that we can move the `actionRegistry.process()` deeper, ie to `onLaunchFinished`, but it still would have to be called before sanity `actionsRegistry.hasPendingAction()` and I see a little difference in bounding this call to either `onLaunchFinished` or `onSwitchConfirmed` - I can move it if you insist, let me know.

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp [hunk 2]
```diff
@@ -367,6 +367,10 @@ namespace app::manager
 
         auto currentlyFocusedApp = getFocusedApplication();
         if (currentlyFocusedApp == nullptr) {
+            if (app->state() != ApplicationCommon::State::NONE) {
+                LOG_INFO("No focused application at the moment, but requested app is starting already...");
+                return false;
```
[the remark this answers is not in the document]
Comment (paweljonskim) [reply module-services/service-appmgr/model/ApplicationManagerCommon.cpp:372]: Im not sure about anything in this system, but static analysis and some tests on the hw (Harmony) did not show anything suspicious. The only part we are throwing out in this check is `onApplicationSwitch` and `startApplication` calls, which basically prevent `stack` (variable) clearing&pushing and multiple application launches if there are many actions incoming to inactive application.

File: module-apps/apps-common/popups/Popups.hpp [hunk 3]
```diff
@@ -29,6 +29,7 @@ namespace gui
             PowerOff,
             Reboot,
             BedtimeNotification,
+            DeadBatteryWindow,
```
[the remark this answers is not in the document]
Comment (paweljonskim) [reply module-apps/apps-common/popups/Popups.hpp:32]: I do agree that its confusing. Adding that to our "technical improvements" site next to AppWindow and build system splits :)

[the remark this answers is not in the document]
Comment (pholat) [reply module-apps/apps-common/popups/Popups.hpp:32]: is it that hard to split though? I believe this would mean right now:
1. moving enum to separate file
2. have separate file for bell and separate for pure 
job done ;)

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pholat) [reply module-services/service-appmgr/model/ApplicationManagerCommon.cpp:372]: this doesn't seem like enough. I understand that this stable is meant for bell right now? Or we will merge it to master and affect pure too?

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pholat) [reply module-services/service-appmgr/model/ApplicationManagerCommon.cpp:720]: I would at least insist to keep the mess in one place - therefore attach it to the function mentioned :)

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (paweljonskim) [reply module-services/service-appmgr/model/ApplicationManagerCommon.cpp:372]: We are planning to merge it to master soon (acutally we are already working on the merge).
I will check if we can live without that change

File: module-apps/apps-common/popups/Popups.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (swichu91) [reply module-apps/apps-common/popups/Popups.hpp:32]: I believe taking such an approach would result in both files having most of their contents identical. I wonder if using the enum class is applicable in this scenario. Maybe it would be better to use templated struct for storing popups types. This way product-specific popups types could be injected at compile time. What do you think @pholat?

[the remark this answers is not in the document]
Comment (pholat) [reply module-apps/apps-common/popups/Popups.hpp:32]: @swichu91 actually this enum doesn't make too much sense in grand scheme of things :) We most probably should use the manifest template to state what popups are in the selected application. It's meant to have such information. Saying that - it's a major update to the code. At the moment I would be happy with just splitting these.

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pholat) [reply module-services/service-appmgr/model/ApplicationManagerCommon.cpp:372]: We are after the merge, so how is it going? :)

File: module-services/service-appmgr/model/ApplicationManagerCommon.cpp [hunk 4]
```diff
@@ -750,8 +754,9 @@ namespace app::manager
             actionsRegistry.finished();
             break;
         default: {
-            auto &params = action->params;
-            app::ApplicationCommon::requestAction(this, app.name(), action->actionId, std::move(params));
+            // IMO this is not neccessary! Commenting for now:
+            // auto &params = action->params;
+            // app::ApplicationCommon::requestAction(this, app.name(), action->actionId, std::move(params));
```
Comment (pholat) [inline module-services/service-appmgr/model/ApplicationManagerCommon.cpp:759]: you do not need to store code in comments in history - you are sure things can be removed? remove them. comments do not compile

**Comment by paweljonskim** [comment]
No longer needed - fixed in other PRs
